sys/arm64/include/asm.h

```diff
@@ -57,19 +57,39 @@
 #define	PIC_SYM(x,y)	x
 #endif
 
-/* Alias for link register x30 */
+#define	INT_WIDTH	8
+#define	INTN(n)		n
+#define	INT(n)		x ## n
+
+#if __has_feature(capabilities)
+#define	CAP_WIDTH	16
+#define	CAPN(n)		c ## n
+#define	CAP(n)		c ## n
+#else
+#define	CAP_WIDTH	INT_WIDTH
+#define	CAPN(n)		INTN(n)
+#define	CAP(n)		INT(n)
+#endif
+
 #ifdef __CHERI_PURE_CAPABILITY__
-#define	REG_WIDTH	16
-#define	REGN(n)		c ## n
-#define	REG(n)		c ## n
+#define	PTR_WIDTH	CAP_WIDTH
+#define	PTRN(n)		CAPN(n)
+#define	PTR(n)		CAP(n)
+/* Alias for link register c30 */
 #define	clr		c30
 #else
-#define	REG_WIDTH	8
-#define	REGN(n)		n
-#define	REG(n)		x ## n
+#define	PTR_WIDTH	INT_WIDTH
+#define	PTRN(n)		INTN(n)
+#define	PTR(n)		INT(n)
+/* Alias for link register x30 */
 #define	lr		x30
 #endif
 
+/* TODO: Remove these confusing deprecated aliases. */
+#define	REG_WIDTH	PTR_WIDTH
```

Comment:
Thanks, I much prefer using PTR/INT/CAP.

Reply:
OTOH, a jump buffer, e.g., needs to save registers (whatever size they are) rather than pointers, per se.  The same is arguably true of argument registers in _rtld_bind.  It's the fact that the full register has to be saved/restored that matters rather than the fact that they are pointers.  It is true that REG generally means GPR and that floating point / vector registers end up with some other prefix/name (e.g. 'struct reg').

This does argue for register_t being a uintcap_t for example, or for the (eventual) goal of 'struct reg' being 'struct capreg' for the purecap userland ABI and only having a 'struct capreg' or NT_CAPREG for the hybrid ABI.  I think all of those are probably sensible things to do though on CHERI architectures with merged register files.  On a split register file architecture, I'd probably say 'struct reg' should grow to contain both for the purecap/native ABI, and you'd probably still want register_t to be a uintcap_t.

Reply:
The name is just too ambiguous in hybrid to be helpful; in my mind REG should be CAP (which is what the kernel generally wants), but in this case it was PTR to match what setjmp wanted.